doc/locale_api.md

```diff
@@ -42,20 +42,123 @@ org.freedesktop.locale1 interface -     -
+`("cz", "qwerty")` or `("us", "basic")`.
+
+We don't expose the console keyboard, instead letting systemd do it via the
+_convert_ parameter.
+
+(The other systemd keyboard settings are X11Model and X11Options, we don't
+have UI or data for that)
+
+```
+# this is gdbus syntax BTW
+node ...Agama/Locale1 {
+  interface ...Agama.Locale1 {
+    methods:
+      # FIXME: the method part is unclear to me. anyone better at proposals?
+      Commit();
+    properties:
+      # NOTE: "as" has different meaning to systemd,
+      # we have a list of LANG settings, 1st gets passed to systemd,
+      # others affect package selection
+      readwrite as   Locale = ['cs_CZ.UTF-8', 'de_DE.UTF-8'];
+      readonly  (ss) X11Keyboard = ('cz','qwerty);
+  };
+};
+```
+
+##### Overriding the User's Choice?
+
+<details>
+<summary>
+If setting the Locale proposes the Language, what do we do if the user first
+changes the language and _then_ the locale?
+</summary>
+
+When Agama UI first shows up, it may show default choices like:
+
+>  Locale: English (US), Keyboard: US
+
+Then we change the locale to Czech, and the keyboard is adjusted automatically:
+
+>  Locale: Czech, Keyboard: Czech
+
+We tune the keyboard:
+
+>  Locale: Czech, Keyboard: Czech (qwerty)
+
+When we then change the locale, the keyboard could stay the same, as we have
+already touched it:
+
+>  Locale: German, Keyboard: Czech (qwerty)
+</details>
+
+##### Simple Design: Always Repropose
+
+We can easily afford throwing away the user's choice of keyboard layout and
+simply set what we consider a good default for a newly set locale, because:
+
+1. it is just one setting (as opposed to whole partitioning layout)
+2. the change will be visible in the UI, I assume
+
+##### Detailed Design: Prioritize
+
+But other cases may not be as simple, so here's a generic design:
+
+All settings are wrapped in a `Priority<T>` generic type (an Enum in Rust),
+meaning, what is the source and importance of the setting:
+- `Machine(data)` means the system has proposed it
+- `Human(data)` means the user has made the choice
+
+In D-Bus, it is represented by wrapping the data in a struct, with a leading
+byte* tagging the priority. For ease of recognition when watching bus traffic,
+special numbers are used:
+- `23` means Human, for the number of chromosome pairs
+- `42` means Machine, as the famous Answer was given by Deep Thought, a machine
+
+In the following dump, we see that the locale was set by the user and the
+system has adjusted the keyboard.
+
+```
+node ...Agama/Locale1 {
+  interface ...Agama.Locale1 {
+    properties:
+      readwrite (yas)   Locale = (23, ['cs_CZ.UTF-8', 'de_DE.UTF-8']);
+      readonly  (y(ss)) X11Keyboard = (42, ('cz','qwerty));
+  };
+};
+```
+
+You may know a [similar settings in libzypp][resstatus] where it has 4 levels.
+
+*: maybe this is a crazy optimization? I am not too opposed to use strings for
+this on the bus.
```

Comment:
or maybe other way. Just use priority like it is on libzypp repos? Another thing to consider here is that beside user/machine it can be also controlled by product, so e.g. some locale is not available on ALP Micro ( now only en is there ). So it can remove invalid value with high priority.

Reply:
re (2) higher priority than Human?? IMHO a bad idea

Reply:
re (1): **priority** is a confusing concept, but I have found nothing better :frowning_face: 

The basic problem is:
1. in **words**, everyone agrees that when lower priority and higher priority meet, the higher priority wins
2. if you express them in **numbers**
  a. _bigger_ number means bigger=higher priority
  b. lower number means higher priority, because a lower number is higher on the page in an _ordered list_.

Anyway, both my numeric values (23 and 42) and libzypp use the "ordered list" semantic.
So, I don't understand what you mean, actually?

Reply:
man zypper:

>  -p, --priority positive-integer

>    Set priority of the repository. Priority of 1 is the highest, the higher the number the lower the priority. Default priority is 99. Packages from repositories with higher priority will be preferred even in case there is a higher installable version available in the repository with a lower priority.

Reply:
Human always wins. (Haven't you watched Terminator? Haven't you known Asimov's laws since you've been ~~manufactured~~ born??)

Product specific? Then specify how it should work and I will name it.